File: hooks/block-git-add-force-staging.py
#!/usr/bin/env python3
"""PreToolUse hook: block git add -f/--force."""

import json
import re
import sys


def main() -> None:
    data = json.load(sys.stdin)
    command = data.get("tool_input", {}).get("command", "")

    # Remove quoted strings to avoid false positives from commit messages etc.
    stripped = re.sub(r'"[^"]*"', '""', command)
    stripped = re.sub(r"'[^']*'", "''", stripped)

    if re.search(r"\bgit\s+add\b", stripped) and re.search(r"\s(-f|--force)\b", stripped):
        json.dump(
            {
                "decision": "block",
                "reason": "git add -f (--force) は禁止されています。.gitignore のルールを迂回する強制ステージングは許可されていません。",
            },
            sys.stdout,
        )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_git_add_force_staging.rs
use regex::Regex;

use crate::input::HookInput;
use crate::output::{block, pass};

pub fn run(input: &HookInput) {
    let command = &input.tool_input.command;
    if command.is_empty() {
        return;
    }

    // 引用符内を除去してフラグ検出の誤検知を防ぐ
    let stripped = Regex::new(r#""[^"]*""#)
        .unwrap()
        .replace_all(command, r#""""#);
    let stripped = Regex::new(r"'[^']*'").unwrap().replace_all(&stripped, "''");

    let git_add_re = Regex::new(r"\bgit\s+add\b").unwrap();
    let force_re = Regex::new(r"\s(-f|--force)\b").unwrap();

    if git_add_re.is_match(&stripped) && force_re.is_match(&stripped) {
        block(
            "git add -f (--force) は禁止されています。.gitignore のルールを迂回する強制ステージングは許可されていません。",
        );
        return;
    }

    pass();
}
